packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,278 @@
+        return this._virtualizedOutbound;
+    }
+
+    constructor(
+        serviceProvider: () => IDocumentService | undefined,
+        logger: ITelemetryLogger,
+        active: () => boolean,
+        createConnectionManager: (props: IConnectionManagerFactoryArgs) => TConnectionManager,
+    ) {
+        super(serviceProvider, logger, active, createConnectionManager);
+        this._virtualizedInbound = new VirtualDeltaQueueProxy(
+            this._inbound,
+            (message: ISequencedDocumentMessage, storeValue?: boolean) => {
+                return this.virtualizeSequencedMessage(message, storeValue);
+            });
+        this._virtualizedOutbound = new VirtualDeltaQueueProxy(
+            this.connectionManager.outbound,
+            (messages: IDocumentMessage[], _storeValue?: boolean) => {
+                return messages.map((message) => {
+                    return {
+                        ...message,
+                        referenceSequenceNumber: this.virtualizeSequenceNumber(message.referenceSequenceNumber),
+                        clientSequenceNumber: this.virtualizeClientSequenceNumber(message.clientSequenceNumber),
+                    }
+                });
+            });
+    }
+
+    public async attachOpHandler(
+        minSequenceNumber: number,
+        sequenceNumber: number,
+        term: number,
+        handler: IDeltaHandlerStrategy,
+        prefetchType: "cached" | "all" | "none" = "none",
+    ) {
+        this._virtualSequenceNumber = sequenceNumber;
+
+        this._sequenceNumberMap.set(sequenceNumber, sequenceNumber);
+        this._sequenceNumberMap.set(minSequenceNumber, minSequenceNumber);
+
+        return super.attachOpHandler(minSequenceNumber, sequenceNumber, term, handler, prefetchType);
+    }
+
+    protected connectHandler(connection: IConnectionDetails) {
+        // TODO: what else do we need to track here?
+        this._clientSequenceNumberMap.clear();
+        this._clientSequenceNumber = 0;
+        super.connectHandler(connection);
+    }
+
+    protected sendMessageToConnectionManager(messages: IDocumentMessage[]): void {
+        if (messages.length >= 2) {
+            this.connectionManager.sendMessages([{
+                type: GroupedBatchOpType,
+                clientSequenceNumber: ++this._clientSequenceNumber,
+                referenceSequenceNumber: messages[0].referenceSequenceNumber, // use oldest referenceSequenceNumber
+                contents: messages,
+            }]);
+        } else {
+            this.connectionManager.sendMessages(messages.map((it: IDocumentMessage) => {
+                this._clientSequenceNumberMap.set(++this._clientSequenceNumber, it.clientSequenceNumber);
+                return {...it, clientSequenceNumber: this._clientSequenceNumber};
+            }));
+        }
+    }
+
+    protected processMessage(message: ISequencedDocumentMessage, startTime: number): void {
+        if (message.type === GroupedBatchOpType) {
```

Comment:
Maybe I'm wrong, but I feel like you are doing double-virtualization - first one when op is pushed into inbound queue, and second one when it is processed here.

Reply:
Same note as above
> The _inbound queue stores the un-virtualized values. It's only when other places try to ask for those values where they get virtualized. All the values are un-virtualized in DeltaManager